### PR TITLE
Icds ucr

### PR DESCRIPTION
**Pull request by sravfeyn**
Starting ICDS custom UCRs. See the [README](https://github.com/dimagi/commcare-hq/pull/10466/files#diff-93e5f7c7e8e78e116988b2d9ac71e939R1) for more context

I am planning to split the code into three PRs involving these three parts. 
1. Custom UCR expressions to make custom indicator calculations possible (with integration tests)
2. Code for custom calculations
3. Tests for these custom calculations

1 allows Sheel (or me) to start writing UCRs. 2 and 3 should ideally be done in one go, but thinking to do it separately - since writing tests for custom business calculations requires stable specs. Getting 2 out will let us use those calculations in UCRs and find more bugs, which in turn can be added in automated tests (3) in subsequent iterations

This PR is for 1st part of above and is good for review

cc @snopoke 
FYI @czue

### CODE REVIEW

File: custom/icds/tests/test_custom_expressions.py [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+                },
+                u'property_value': 'child'
+            },
+            configured_indicators=[
+                {
+                    "type": "expression",
+                    "column_id": "age_in_months",
+                    "datatype": "integer",
+                    "expression": {
+                        'type': 'child_health_property',
```
Comment (sravfeyn) [inline custom/icds/tests/test_custom_expressions.py]: @sheelio This is an example structure of how the ICDS UCRs are going to look like. Can you take a look at this UCR config and confirm if this looks good?

File: custom/icds/ucr/expressions.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+    def __call__(self, item, context=None):
+        case_doc = context.root_doc
+
+        start_date = self._start_date(item, context)
+        end_date = self._end_date(item, context)
+
+        child_health_case = ChildHealthCaseRow(case_doc, transform_date(start_date), transform_date(end_date))
+
+        return getattr(child_health_case, self.indicator_name, None)
```
Comment (sravfeyn) [inline custom/icds/ucr/expressions.py]: @czue Can you review this to see if this looks good and I am not doing anything hackish with the framework?

File: custom/icds/models.py [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+            form.get_data("form/z_score_wfa_ql/nutrition_status")
+            for form in self.filtered_forms(GMP_XMLNS)
+            if form.get_data("form/z_score_wfa_ql/nutrition_status")
+        ]
+        return statuses[0] if any(statuses) else 'unweighed'
+
+    def dob(self):
+        return transform_date(self.mother_case.case_property("dob"))
+
+    def self.age_in_days(self):
```
Comment (dimagimon) [inline custom/icds/models.py]: E901 SyntaxError: invalid syntax

File: custom/icds/models.py [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+        return one_or_zero(self.nutrition_status == 'normal')
+
+    def nutrition_status_unweighed(self):
+        return one_or_zero(self.nutrition_status == 'unweighed')
+
+    def nutrition_status_weighed(self):
+        return one_or_zero(self.nutrition_status != 'unweighed')
+
+    def num_rations(self):
+        gmp_forms = self.filtered_forms(GMP_XMLNS)
```
Comment (dimagimon) [inline custom/icds/models.py]: F841 local variable 'gmp_forms' is assigned to but never used

File: custom/icds/models.py [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+
+    def thr_eligible(self):
+        return one_or_zero(6 <= self.age_in_months <= 36)
+
+    def pse_eligible(self):
+        return one_or_zero(self.age_in_months > 36)
+
+    def pse_days_attended(self):
+        pse_forms = self.filtered_forms(PSE_XMLNS)
```
Comment (dimagimon) [inline custom/icds/models.py]: F841 local variable 'pse_forms' is assigned to but never used

**Comment by sheelio** [comment]
@sravfeyn I think this looks good - I'm guessing cleaner to have one expression with a parameter than multiple parameters?

File: custom/icds/ucr/README.md [hunk 6]
```diff
@@ -0,0 +1,26 @@
+## Custom UCR expressions for ICDS reports
+
+### Reports background
+ICDS Tableau reporting uses UCR data-sources as the base data. The UCR datasources are mainly case data and they are configured to calculate indicators as on a snapshot duration (for e.g. a child case as a year ago) (this is done via stock UCR using `base_item_expression`). Most of these indicators need form lookups which can't be done in stock UCR. Following custom UCR expressions are made available to make such form lookup calculations possible in ICDS reports.
+
+### Custom UCR expressions
```
Comment (sravfeyn) [inline custom/icds/ucr/README.md]: @sheelio FYI - docs

File: custom/icds/models.py [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+        dates can be date/datetime objects or a valid strings
+        """
+        if not isinstance(case, CommCareCase):
+            case = CommCareCase.wrap(case)
+        self.case = case
+        self.snapshot_start_date = transform_date(snapshot_start_date)
+        self.snapshot_end_date = transform_date(snapshot_end_date)
+        self.forms_provider = forms_provider
+
+    # needs caching
```
Comment (sravfeyn) [inline custom/icds/models.py]: Still figuring out this and planning to do this in a separate PR. Safe to not block on this, as these wont be used in the UCRs yet.

File: custom/icds/tests/test_custom_expressions.py [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+                },
+                u'property_value': 'child'
+            },
+            configured_indicators=[
+                {
+                    "type": "expression",
+                    "column_id": "age_in_months",
+                    "datatype": "integer",
+                    "expression": {
+                        'type': 'child_health_indicator',
```
Comment (sravfeyn) [inline custom/icds/tests/test_custom_expressions.py]: As @sheelio pointed out, we can have one custom expression for all case types, instead of one for each case-type. In fact, I prefer to have one custom expression (since that will be more DRY in `ChildHealthIndicatorExpressionSpec` as well).

File: custom/icds/models.py [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+    def forms(self):
+        if self.forms_provider:
+            return self.forms_provider
+        forms = []
+        for action in self.case.actions:
```
Comment (snopoke) [inline custom/icds/models.py]: Should use `case.xform_ids` and then `FormAccessors(domain).get_form(form_id` to make it compatible with SQL models

File: custom/icds/models.py [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+            forms.append(XFormInstance.get(action.xform_id))
+
+    def case_property(self, case, property, default=None):
+        prop = getattr(case, property, default)
```
Comment (snopoke) [inline custom/icds/models.py]: If this is always going to be dynamic properties then use `case.dynamic_case_properties().get(prop, default)`

File: custom/icds/models.py [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+
+    def filtered_forms(self, xmlns):
+        for form in self.forms:
+            if form.xmlns == xmlns and are_in_same_month(form.received_on, self.snapshot_start_date):
+                yield form
+
+    def is_open(self):
+        return are_in_same_month(self.case.opened_on, self.snapshot_start_date)
+
+    def nutrition_status(self):
```
Comment (snopoke) [inline custom/icds/models.py]: A number of these should be `memoized` to avoid computing repeatedly

File: custom/icds/models.py [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+
+        args:
+            case: case of type 'child_health'
+            snapshot_start_date: Start date of duration to snapshot the case on
+            snapshot_end_date: End date of duration to snapshot the case on.
+
+        dates can be date/datetime objects or a valid strings
+        """
+        if not isinstance(case, CommCareCase):
+            case = CommCareCase.wrap(case)
```
Comment (snopoke) [inline custom/icds/models.py]: could be troublesome with SQL models

Comment (snopoke) [inline custom/icds/models.py]: Not sure the best solution at the moment

File: custom/icds/ucr/expressions.py [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+        self._start_date = start_date
+        self._end_date = end_date
+
+    def __call__(self, item, context=None):
+        start_date = self._start_date(item, context)
+        end_date = self._end_date(item, context)
+
+        case_model = CASE_TYPE_TO_MODEL_MAPPING.get(self.case_type, None)
+        case_doc = context.root_doc
+        child_health_case = case_model(case_doc, transform_date(start_date), transform_date(end_date))
```
Comment (snopoke) [inline custom/icds/ucr/expressions.py]: Can an expression add things to the context as a temporary cache?

Comment (sravfeyn) [inline custom/icds/ucr/expressions.py]: Not sure in general, but at least here we cant add things to the the context kwarg. It is a `EvaluationContext` object (not a generic context dict). What do you have in mind ?

File: custom/icds/models.py (hunk 12 above)
Comment (sravfeyn) [inline custom/icds/models.py]: Isn't most of UCR built around the fact that the data is in json-format? What are the implications of SQL models on UCR overall? Is there doc I should be looking at, to write SQL-proof code?

**Comment by sravfeyn** [comment]
@snopoke , @czue Would be good to get some feedback and have this merged

File: custom/icds/models.py (hunk 12 above)
Comment (snopoke) [inline custom/icds/models.py]: Yes, UCR is build around the assumption that the data is a `dict

**Comment by czue** [comment]
taking a look at this now. fyi, still has a "don't merge" label on it - not sure if that's intentional

**Comment by sravfeyn** [comment]
(Kept the label in case there was gonna be considerable feedback on this)

File: corehq/apps/userreports/expressions/__init__.py [hunk 14]
```diff
@@ -2,6 +2,15 @@
...
 from corehq.apps.userreports.expressions.factory import ExpressionFactory
 
+
+def get_custom_ucr_expressions():
+    custom_ucr_expressions = settings.CUSTOM_UCR_EXPRESSIONS
+
+    for expression_list in settings.CUSTOM_UCR_EXPRESSION_LISTS:
+        custom_ucr_expressions += import_string(expression_list)
+
+    return custom_ucr_expressions
```
Comment (czue) [inline corehq/apps/userreports/expressions/__init__.py]: copying over comment from https://github.com/dimagi/commcare-hq/pull/10467:

is this not calling `import_string` on the whole list? also, even if that works won't the first half of the list be string tuples and the second half be imported modules? finally, if `CUSTOM_UCR_EXPRESSIONS` is a list and we add to it won't that get added over and over again each time?

```python
>>> x = [1, 2, 3]
>>> y = x
>>> y += [4, 5]
>>> x
[1, 2, 3, 4, 5]
>>> y += [4, 5]
>>> x
[1, 2, 3, 4, 5, 4, 5]
>>> 
```

File: custom/icds/constants.py [hunk 15]
```diff
@@ -0,0 +1,4 @@
+GMP_XMLNS = "http://openrosa.org/formdesigner/b183124a25f2a0ceab266e4564d3526199ac4d75"
```
Comment (czue) [inline custom/icds/constants.py]: I think `custom.icds` is missing an `__init__.py`

File: custom/icds/models.py [hunk 16]
```diff
@@ -0,0 +1,163 @@
+
+from casexml.apps.case.models import CommCareCase
+from corehq.apps.userreports.expressions.getters import transform_date
+
+from dimagi.utils.decorators.memoized import memoized
+
+from .constants import *
```
Comment (czue) [inline custom/icds/models.py]: nit but I prefer `from . import constants` and then referencing things as `constants.FOO`. avoids having an `import *`

File: custom/icds/models.py [hunk 17]
```diff
@@ -0,0 +1,163 @@
...
+
+
+class ChildHealthCaseRow(object):
+    def __init__(self, case, snapshot_start_date, snapshot_end_date, forms_provider=[]):
```
Comment (czue) [inline custom/icds/models.py]: shouldn't have mutable default argumantes

File: custom/icds/models.py (hunk 12 above)
Comment (czue) [inline custom/icds/models.py]: > If possible it would be best to leave the case as JSON.

agree with this. if this isn't possible we should minimize the number of places we make the couch assumption and abstract them out so we can plug in sql adapters later.

File: custom/icds/models.py [hunk 18]
```diff
@@ -0,0 +1,163 @@
...
+        """
+        Snapshot of the case in the duration specified by snapshot_start_date and snapshot_end_date
+
+        args:
+            case: case of type 'child_health'
+            snapshot_start_date: Start date of duration to snapshot the case on
+            snapshot_end_date: End date of duration to snapshot the case on.
+
+        dates can be date/datetime objects or a valid strings
```
Comment (czue) [inline custom/icds/models.py]: was there a reason you chose to allow both of these defaults? i realize typing in UCR is messy, but just curious

File: custom/icds/models.py [hunk 19]
```diff
@@ -0,0 +1,163 @@
...
+
+class ChildHealthCaseRow(object):
+    def __init__(self, case, snapshot_start_date, snapshot_end_date, forms_provider=[]):
+        """
+        Snapshot of the case in the duration specified by snapshot_start_date and snapshot_end_date
+
+        args:
+            case: case of type 'child_health'
+            snapshot_start_date: Start date of duration to snapshot the case on
+            snapshot_end_date: End date of duration to snapshot the case on.
```
Comment (czue) [inline custom/icds/models.py]: might be good to document `forms_provider`. it seems like the hardest to understand argument

File: custom/icds/models.py [hunk 20]
```diff
@@ -0,0 +1,163 @@
...
+    def forms(self):
+        if self.forms_provider:
+            return self.forms_provider
+        forms = []
+        for form_id in self.case.xform_ids:
+            forms.append(FormAccessors(domain).get_form(form_id))
+
+    def case_property(self, case, property, default=None):
+        prop = case.dynamic_case_properties().get(property, default)
+        if isinstance(prop, basestring) and prop.strip() == "":
```
Comment (czue) [inline custom/icds/models.py]: nit: this feels like a strange place for this logic. shouldn't `case_property` just return whatever property is specified in the case?

File: custom/icds/models.py [hunk 21]
```diff
@@ -0,0 +1,163 @@
...
+    @memoized
+    def forms(self):
+        if self.forms_provider:
+            return self.forms_provider
+        forms = []
+        for form_id in self.case.xform_ids:
+            forms.append(FormAccessors(domain).get_form(form_id))
+
+    def case_property(self, case, property, default=None):
+        prop = case.dynamic_case_properties().get(property, default)
```
Comment (czue) [inline custom/icds/models.py]: I think this function at least should be easy to rewrite with the dict.

File: custom/icds/models.py [hunk 22]
```diff
@@ -0,0 +1,163 @@
...
+            return default
+        return prop
+
+    def filtered_forms(self, xmlns):
+        for form in self.forms:
+            if form.xmlns == xmlns and are_in_same_month(form.received_on, self.snapshot_start_date):
+                yield form
+
+    @property
+    @memoized
```
Comment (czue) [inline custom/icds/models.py]: not sure memoization is adding much on some of these things that don't look like they're doing much work.

File: custom/icds/models.py [hunk 23]
```diff
@@ -0,0 +1,163 @@
...
+    def mother_case(self):
+        return self.case.parent
+
+    @property
+    @memoized
+    def household_case(self):
+            return self.mother_case.parent if self.mother_case else None
```
Comment (czue) [inline custom/icds/models.py]: nit: weird intentation

File: custom/icds/models.py [hunk 24]
```diff
@@ -0,0 +1,163 @@
...
+
+    def nutrition_status_severely_underweight(self):
+        return one_or_zero(self.nutrition_status == 'severely underweight')
```
Comment (czue) [inline custom/icds/models.py]: it looks like a lot of these could be computed in stock UCR - why are they being added here?

File: custom/icds/models.py (hunk 18 above)
Comment (sravfeyn) [inline custom/icds/models.py]: ICDS has a use case where they will have different business months for different reports

File: custom/icds/constants.py (hunk 15 above)
Comment (sravfeyn) [inline custom/icds/constants.py]: It had one https://github.com/dimagi/commcare-hq/blob/master/custom/icds/__init__.py

**Comment by czue** [comment]
I'm not sure I totally understood how you were planning on architecting this when we talked and I have some concerns about the current setup. Did you consider other options, e.g. making a base set of expressions to get and manipulate forms? Just worried that the current setup could get pretty unmaintainable. This seems to be more like the OPM reports than how I would have expected things to fit into UCR paradigms (though I haven't thought that hard about what the other options might be).

File: custom/icds/constants.py (hunk 15 above)
Comment (czue) [inline custom/icds/constants.py]: :+1: sorry thought it was new

File: custom/icds/ucr/expressions.py (hunk 13 above)
Comment (czue) [inline custom/icds/ucr/expressions.py]: we could definitely add a `cache` property to `EvaluationContext`

**Comment by sravfeyn** [comment]
Here are the type of calculations that I didn't know could be done in stock UCR, which I have done here in custom calculations.

- Calculations involving arithmatic (like age calculation). I imagined writing arithmatic in UCR would either involve some kind of expression parser or a context-free-grammer implementation - either of those two routes seemed high LOE
- Calculations involving form filtering (in specific months). To describe it formally - given a case, a snapshotting month, a form type, a relative duration before snapshotting month and a form-property, return list of properties in forms filtered according to relative duration. It's strightforward to write some kind of expression for this - but the resulting UCR expression that users have to write seems super complicated. Are you imagining more simple things with 'base set of expressions to get and manipulate  forms'?

Other UCR-able Calculations that involve result of one of above calculations (for e.g. age-group based on age) - these can be written in stock UCR, but I have written in custom calculations here, I will remove them

I agree that this is looking lot like OPM architecture (mainly because of the snapshotting logic and both being MCH apps). Plan is to avoid using custom python calculations for everything that can be done in UCR. Also, I intended this PR as only a starting point for architectural discussion (that's why I had the 'dont-merge-label') and am open to change this altogether.

**Comment by sravfeyn** [comment]
(Will respond to other code related PR feedback once we are sure of this approach)

File: corehq/apps/userreports/expressions/__init__.py (hunk 14 above)
Comment (sravfeyn) [inline corehq/apps/userreports/expressions/__init__.py]: Updated #10467

**Comment by sravfeyn** [comment]
Closing this, in favour of making the calculations listed [here](https://github.com/dimagi/commcare-hq/pull/10466#issuecomment-190262931) available in stock UCR
